src/ray/object_manager/pull_manager.cc

```diff
@@ -90,28 +91,31 @@ void PullManager::TryPull(const ObjectID &object_id) {
     return;
   }
 
-  // Choose a random client to pull the object from.
-  // Generate a random index.
-  std::uniform_int_distribution<int> distribution(0, node_vector.size() - 1);
-  int node_index = distribution(gen_);
-  NodeID node_id = node_vector[node_index];
-  // If the object manager somehow ended up choosing itself, choose a different
-  // object manager.
-  if (node_id == self_node_id_) {
-    std::swap(node_vector[node_index], node_vector[node_vector.size() - 1]);
-    node_vector.pop_back();
-    RAY_LOG(WARNING)
-        << "The object manager with ID " << self_node_id_ << " is trying to pull object "
-        << object_id << " but the object table suggests that this object manager "
-        << "already has the object. It is most likely due to memory pressure, object "
-        << "pull has been requested before object location is updated.";
-    node_id = node_vector[node_index % node_vector.size()];
-    RAY_CHECK(node_id != self_node_id_);
+  NodeID to_pull;
+  int num_potential_pulls = 0;
+  for (auto &node_id : node_vector) {
```

Comment:
Might be simpler to sample from client_locations, then move the NodeID from client_locations to inflight.

Reply:
Hmm I think the issue is that `OnLocationsChanged` could be called, leading to the object being in both the inflight and client_locations list.

Reply:
That's why I suggested doing it this way:
> 2. On new locations, reset the current list of locations, but remove each one that appears in the set you've tried so far (and maybe also remove ones that no longer have the object).

But I think this is okay too.